Authority validation strategy

eric-b opened the pull request:
Fixes #226

eric-b committed:
Authority validation strategy

eric-b committed:
Fix doc

eric-b committed:
Removed ncrunch solution files

eric-b committed:
Fix doc

leastprivilege commented on src/Client/AuthorityUrlValidationStrategy.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <summary>
+        /// Singleton instance.
+        /// </summary>
+        public static AuthorityUrlValidationStrategy Instance = new AuthorityUrlValidationStrategy();
```
Is there a good reason to use the "Instance" pattern here?

eric-b replied:
My reasoning about that is that creating different instances brings no benefit over singleton "Instance" pattern (no parameterization and no state). Probably not a "good" reason. I removed that.

leastprivilege commented on src/Client/AuthorityUrlValidationStrategy.cs:
```diff
@@ -0,0 +1,62 @@
+
+        public AuthorityValidationResult IsIssuerNameValid(string issuerName, string expectedAuthority)
+        {
+            if (!Uri.TryCreate(expectedAuthority.RemoveTrailingSlash(), UriKind.Absolute, out var expectedAuthorityUrl))
```
Please use { } here...

leastprivilege commented on src/Client/AuthorityUrlValidationStrategy.cs:
```diff
@@ -0,0 +1,62 @@
+
+            if (string.IsNullOrWhiteSpace(issuerName)) return AuthorityValidationResult.CreateError("Issuer name is missing");
+
+            if (!Uri.TryCreate(issuerName.RemoveTrailingSlash(), UriKind.Absolute, out var issuerUrl))
```
..and here

leastprivilege commented on src/Client/AuthorityUrlValidationStrategy.cs:
```diff
@@ -0,0 +1,62 @@
+            if (!Uri.TryCreate(issuerName.RemoveTrailingSlash(), UriKind.Absolute, out var issuerUrl))
+                return AuthorityValidationResult.CreateError("Issuer name is not a valid URL");
+
+            if (expectedAuthorityUrl.Equals(issuerUrl))
```
..and here

leastprivilege commented on src/Client/AuthorityUrlValidationStrategy.cs:
```diff
@@ -0,0 +1,62 @@
+        public AuthorityValidationResult IsEndpointValid(string endpoint, IEnumerable<string> allowedAuthorities)
+        {
+            if (string.IsNullOrEmpty(endpoint))
+                return AuthorityValidationResult.CreateError("endpoint is empty");
```
IOW - I really don't like omission of curly braces.. I could do it myself - or do you want to change that?

leastprivilege commented on src/Client/DiscoveryPolicy.cs:
```diff
@@ -19,7 +22,25 @@ public class DiscoveryPolicy
+        /// Strategy used to validate issuer name and endpoints based on expected authority.
+        /// Defaults to <see cref="StringComparisonAuthorityValidationStrategy.Default"/>.
+        /// </summary>
+        public IAuthorityValidationStrategy AuthorityValidationStrategy { get; set; } = new StringComparisonAuthorityValidationStrategy(defaultAuthorityNameComparison);
```
Here you are newing up the strategy - instead of using Instance. I'd prefer consistency.

eric-b replied:
I used "Instance" with the non parameterizable class AuthorityUrlValidationStrategy.
The class StringComparisonAuthorityValidationStrategy is parameterizable with StringComparison. IMHO it's consistent, but it maybe is a matter of taste.

eric-b committed:
Changes following PR review

eric-b committed:
- Fixed regression in AuthorityUrlValidationStrategy.
- Updated unit tests to validate the two different strategies for each test depending on authority validation.
- Obsolete stuff removed (breaking changes)
